BHoM_Engine/Query/FindIdentifier.cs

```diff
@@ -0,0 +1,83 @@
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using BH.oM.Base;
+using BH.oM.Reflection.Attributes;
+using System.ComponentModel;
+
+namespace BH.Engine.Base
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Tries to find an IAdapterId on the object.")]
+        [Input("o", "The object to search for an IAdapterId.")]
+        [Output("identifier", "First plausible identifier present on the object.")]
+        public static Type FindIdentifier(this IBHoMObject o)
+        {
+            return o?.Fragments.FirstOrDefault(fr => fr is IAdapterId)?.GetType();
+        }
+
+        /***************************************************/
+
+        [Description("Tries to find a AdapterIdType on the object if no input is provided.")]
+        [Output("adapterIdType", "First plausible identifier present on the object or provided.")]
+        public static Type FindIdentifier(this IBHoMObject o, Type adapterIdType)
+        {
+            if (o == null)
+            {
+                Reflection.Compute.RecordError("The object provided is null, please check inputs.");
+                return null;
+            }
+            if (adapterIdType == null)
+            {
+                adapterIdType = o.FindIdentifier();
+                if (adapterIdType == null)
+                {
+                    Reflection.Compute.RecordError("No Identifier found.");
+                    return null;
+                }
+                else
+                    Reflection.Compute.RecordNote("Auto-generated Identifier as " + adapterIdType.Name);
+            }
+            else if (!typeof(IAdapterId).IsAssignableFrom(adapterIdType))
+            {
+                Reflection.Compute.RecordError("The provided adapterIdType need to be a type of IAdapterId.");
+                return null;
+            }
+
+            return adapterIdType;
```

Comment:
With the above, this can be simplified down to:

```suggestion
            if (adapterIdType == null)
            {
               return o.FindIdentifier();
            }
            else if (!typeof(IAdapterId).IsAssignableFrom(adapterIdType))
            {
                Engine.Reflection.Compute.RecordError("The provided adapterIdType need to be a type of IAdapterId.");
                return null;
            }
            return adapterIdType;
```